# pylibseq: Python bindings for libsequence

This package provides [Python](http://www.python.org/) bindings for the
C++11 library [libsequence](http://molpopgen.github.io/libsequence/).

The bindings are implemented using
[pybind11](http://pybind11.readthedocs.io/).

The original package no longer built in an environment with an updated cmake,
pybind11, and modern Python3 (>3.9) (I'm not sure which caused the build errors).
Anyways, setup.py seems legacy at this point.

This package now builds with scikit-build-core.

This package serves two roles:

-   It provides a means of using some of the more widely-used bits of
    [libsequence](http://molpopgen.github.io/libsequence/) within the
    Python language
-   The unit tests of this package also serve as unit tests for
    [libsequence](http://molpopgen.github.io/libsequence/).

What this package does **not** (currently) do:

-   provide an interface for I/O operations. Python I/O and C++ I/O are
    fundamentally very different. Bridging the gap requires either
    adding features to pybind11 and/or adding modules to this package
    that depend on the [boost](http://www.boost.org/) Python interface,
    which would add an additional C++ dependency to this package.

## Installing

Clone this repository. Then, install the package with:

```sh
cd pylibseq
pip install .
```

I'll setup CI to build wheels for PyPi soon.

I haven't tested, but the conda install for the now-archived original repository is:

``` bash
conda -c bioconda install pylibseq
```

Verify your installation works with:

```python
import pylibseq # CHANGED! was 'import libsequence'

data = pylibseq.SimData(x)
f = pylibseq.Fst(self.d,[2,2])
print(f)
```

## Citation

If you use this software for your research, please cite:

    @ARTICLE{Thornton2003-wj,
      title    = "Libsequence: a C++ class library for evolutionary genetic
                  analysis",
      author   = "Thornton, Kevin",
      abstract = "UNLABELLED: A C++ class library is available to facilitate the
                  implementation of software for genomics and sequence polymorphism
                  analysis. The library implements methods for data manipulation
                  and the calculation of several statistics commonly used to
                  analyze SNP data. The object-oriented design of the library is
                  intended to be extensible, allowing users to design custom
                  classes for their own needs. In addition, routines are provided
                  to process samples generated by a widely used coalescent
                  simulation. AVAILABILITY: The source code (in C++) is available
                  from http://www.molpopgen.org",
      journal  = "Bioinformatics",
      volume   =  19,
      number   =  17,
      pages    = "2325--2327",
      month    =  nov,
      year     =  2003,
      url      = "https://www.ncbi.nlm.nih.gov/pubmed/14630667",
      language = "en",
      issn     = "1367-4803",
      pmid     = "14630667",
      doi      = "10.1093/bioinformatics/btg316"
    }

## Requirements:

-   python 3
-   a modern c++ compiler
-   cmake \>= 3.18

## Changelog (rough)

-   0.3.0 (new maintainer): package now uses scikit-build-core as the
    build backend.
-   0.2.0: The package has been completely refactored. We now use
    [pybind11](http://pybind11.readthedocs.io/) to integrate C++ and
    Python. Previous versions of this project used
    [Cython](http://www.cython.org). The API now corrresponds to
    [libsequence](http://molpopgen.github.io/libsequence/) 1.9.2.
    [Python](http://www.python.org/) \>= 3.4 is required.
-   0.1.9: Made memory management more robust: more unique_ptr instead
    of raw pointers. Cleanup \_\_dealloc\_\_ functions in extension
    types. Package now sets \_\_version\_\_. Class names are now
    \"Pythonic\" (and identical to the corresponding type names from
    libsequence) due to aliasing the C++ names from libsequence. Change
    from distutils to setuptools. Documentation fixes. Expose haplotype
    diversity and number of haplotype statistics. First (very alpha)
    release of pymsstats.
-   0.1.8: made sure C++ objects/fxns are declared \"nogil\". Raw
    pointers replaced with C++\'s unique_ptr.
-   0.1.7: improvements to build system. Add option to build from
    GitHub.
-   0.1.6: update to libsequence 1.8.9. Add \--use-cython option to
    setup.py


## Documentation:

-   Project [homepage](http://molpopgen.github.io/pylibseq/)
-   [Manual](http://molpopgen.github.io/pylibseq/docs/_build/html/index.html)
